### ISSUE TITLE
Light cache is too limited for open space and prop buildings

### ISSUE THREAD

**Issue by grinchfox** [Facepunch/garrysmod-issues#2815]
### Details

Since Garry's Mod has high entity limit it is still struggling displaying more than about 200 lit entities placed in different positions. Continuous light update causes massive lag on client side.
I think the limit should be more adjustable to handle big maps or just increased.

![image_2016-09-10_18-25-53](https://cloud.githubusercontent.com/assets/3694039/18411324/c07b7242-777d-11e6-986b-810f3abb4610.png)
Here is also [example video](https://vixen.nxserv.eu/owncloud/index.php/s/uamD4Qc3CAKLCtJ/download) of  what happens when engine is trying to fit the light cache.

As I understand max light cache entries are defined [here (source2007)](https://github.com/LestaD/SourceEngine2007/blob/43a5c90a5ada1e69ca044595383be67f40b33c61/src_main/engine/lightcache.cpp#L56)
### Steps to reproduce

Spawn more than 200 props somewhere around, these origins must be at least about 64 units far from each other.
Then you may see how game tries to relit all entities but doesn't fits cache so some of them blinking as shown on video above. You may also see `Light Cache` graph high in budget profiler as on screenshot above.

**Comment by VigiFyre** [comment]
[example video of what happens when engine is trying to fit the light cache](http://nxserv.eu/static/2016-09-10_19-10-31.mp4)

**Comment by Noiwex** [comment]
This is a real issue. @robotboy655 @willox

**Comment by mcNuggets1** [comment]
True problem.

**Comment by VigiFyre** [comment]
[a better demonstration of lightcache performance impact](https://nxserv.eu/static/2016-10-21_21-43-43.mp4)

**Comment by robotboy655** [comment]
I will try to increase it after next weeks hotfix to like 3000 in Dev, to see if it actually makes any difference.

**Comment by Noiwex** [comment]
![](http://i.giphy.com/NnGGHE0muVqpO.gif)

**Comment by robotboy655** [comment]
It's on Dev, please test and report if there are any improvements.

**Comment by uRandomAlex** [comment]
@robotboy655 Crashes after joining a game and playing for some time (no more than 10 minutes).

**Comment by VigiFyre** [comment]
It seems to have worked.

Same scene:

live branch
![](http://i.imgur.com/JzUuW0d.jpg)

dev branch
![](http://i.imgur.com/TbYQ29t.jpg)

**Comment by robotboy655** [comment]
Assuming being solved.

**Comment by ZehM4tt** [comment]
It seems you haven't corrected some other things.

```
// this should be prime to make the hash better
#define MAX_CACHE_ENTRY		3001
#define MAX_CACHE_BUCKETS	MAX_CACHE_ENTRY

// number of bits per grid in x, y, z
#define HASH_GRID_SIZEX		10
#define	HASH_GRID_SIZEY		10
#define	HASH_GRID_SIZEZ		13
```


This should be working. 

Note those HASH_GRID_SIZE constants, 5 bits would not work for the 3000 entries and 3000 isn't a prime number.

**Comment by robotboy655** [comment]
200 isn't a prime number either, yet it was the old value. I assumed the prime comment was related to the hash grid size.

My HASH_GRID_SIZE values are a bit bigger to be closer to 3000 (x*y*z) while being prime numbers. 

You numbers fix the hl2 lighting issue, but I am not sure how it will reflect on big map/lots-o-prop performance.

**Comment by Kefta** [comment]
What is breaking the HL2 lighting exactly with the lightcache raising?

**Comment by robotboy655** [comment]
Actually even your values break the lighting in other parts on the same map

**Comment by Noiwex** [comment]
Try 20/20/26?

**Comment by Kefta** [comment]
Those aren't prime numbers

**Comment by robotboy655** [comment]
Doesn't make a difference, all NPCs are improperly lit on d1_trainstation_02 now.

![image](https://cloud.githubusercontent.com/assets/3299036/23135213/645f02a0-f7a0-11e6-9b0e-3841a048feb5.png)

**Comment by ZehM4tt** [comment]
Will do some more tests, the whole origin to cache origin seems awful.

**Comment by PotcFdk** [comment]
Help me understand the issue here, what happened to the "goold old" XORing the components with large primes and calculating the 0 <= x < table size?

**Comment by ZehM4tt** [comment]
I made the entire thing dynamic after all and replaced the hash with something better fit for 3d coordinates. I have now r_lightcachesize, r_lightcachegridx, r_lightcachegridy, r_lightcachegridz, r_lightcachesnapepsilon.

![hl2_2017-02-21_05-24-51](https://cloud.githubusercontent.com/assets/5415177/23150735/18d9fc9e-f7f6-11e6-856f-ef4056f0e1c3.jpg)

Rubat if you are interested into how I did it contact me on Steam.
